user/tests/wiring/power_management_long_running/power.cpp

```diff
@@ -0,0 +1,379 @@
+        return System.batteryState() == BATTERY_STATE_DISCONNECTED;
+    }, 10000));
+    assertTrue(System.powerSource() == POWER_SOURCE_USB_HOST);
+
+    SERIAL.println("> Test is running...");
+    time32_t start = Time.now();
+    while ((Time.now() - start) < TEST_DURATION) {
+        assertTrue(System.batteryState() == BATTERY_STATE_DISCONNECTED);
+        assertTrue(System.powerSource() == POWER_SOURCE_USB_HOST);
+        delay(1s);
+    }
+
+    g_state = DISCONNECTED_STATE_SLEEP;
+}
+
+test(power_07_battery_disconnected_and_sleep) {
+    SERIAL.println("\r\npower_07_battery_disconnected_and_sleep");
+    if (g_state != DISCONNECTED_STATE_SLEEP) {
+        skip();
+        return;
+    }
+
+    constexpr system_tick_t SLEEP_DURATION = 10 * 1000; // 10 seconds
+
+    SERIAL.printlnf("> Device will sleep for %ld seconds", SLEEP_DURATION);
+    SystemSleepResult r = System.sleep(SystemSleepConfiguration().mode(SystemSleepMode::STOP).duration(SLEEP_DURATION));
+    assertEqual(r.error(), SYSTEM_ERROR_NONE);
+    assertTrue(r.wakeupReason() == SystemSleepWakeupReason::BY_RTC);
+
+    SERIAL.println("> Device wakes up. Please reconnect the Serial to continue the test.");
+    while (!Serial.isConnected());
+    delay(1s);
+
+    SERIAL.println("> Wait the power management settled...");
+    assertTrue(waitFor([]{
+        return System.batteryState() == BATTERY_STATE_DISCONNECTED;
+    }, 10000));
+    assertTrue(System.powerSource() == POWER_SOURCE_USB_HOST);
+
+    g_state = CONNECTED_STATE_SLEEP;
+}
+
+test(power_08_battery_connected_and_sleep) {
+    SERIAL.println("\r\npower_08_battery_connected_and_sleep");
+    if (g_state != CONNECTED_STATE_SLEEP) {
+        skip();
+        return;
+    }
+
+    constexpr system_tick_t SLEEP_DURATION = 10 * 1000; // 10 seconds
+
+    SERIAL.println("> Please \"plug\" the battery. Press any key to continue when the hardware is set up.");
+    while (SERIAL.available() <= 0);
+    while (SERIAL.available() > 0) {
+        (void)SERIAL.read();
+    }
+
+    SERIAL.println("> Wait the power management settled...");
+    assertTrue(waitFor([]{
+        return System.batteryState() != BATTERY_STATE_UNKNOWN;
+    }, 10000));
+    assertTrue(System.powerSource() == POWER_SOURCE_USB_HOST);
+
+    delay(10s); // When battery is pluged, the PMIC may take some time to finally settle the battery state.
+    static int currState = System.batteryState();
+
+    SERIAL.printlnf("> Device will sleep for %ld seconds", SLEEP_DURATION);
+    SystemSleepResult r = System.sleep(SystemSleepConfiguration().mode(SystemSleepMode::STOP).duration(SLEEP_DURATION));
+    assertEqual(r.error(), SYSTEM_ERROR_NONE);
+    assertTrue(r.wakeupReason() == SystemSleepWakeupReason::BY_RTC);
+
+    SERIAL.println("> Device wakes up. Please reconnect the Serial to continue the test.");
+    while (!Serial.isConnected());
+    delay(1s);
+
+    SERIAL.println("> Wait the power management settled...");
+    assertTrue(waitFor([&]{
```

Comment:
This one also failed on a Boron and again it's difficult to see what the problem was:
```
Assertion failed: (System.waitCondition([]{ return ([&]{ return System.batteryState() == currState; })(); }, (10000))=0) == (true=1), file tests/wiring/power_management_long_running/power.cpp, line 375.
```

The device was blinking charge LED in a pattern of `.-.-.-.-.-.-.-.----------------------------------.-.-.-.-.-.-.-.-.-----------------------------` for a couple of minutes after which the charge LED settled on full-on (at this point the test has already failed).

Reply:
This happened close to termination voltage, so I assume most likely our hack with re-charge threshold may not be working entirely correctly or we cannot do anything about this potentially and simply need to account for this behavior in the test.

Reply:
This symptom is probably caused by the recharged threshold and happened on part of batteries. Unless we keep the 100mV recharge threshold all the time, we may see this occasionally.